### ISSUE TITLE
Simple stencil with vertical offset fails for gtc:gt backends

### ISSUE THREAD

**Issue by tehrengruber** [GridTools/gt4py#437]
The following stencils fails with the new `gtc:gt` backends, e.g. `gtc:gt:cpu_ifirst` during dace oir optimization.

```python
@stencil(backend=backend, **backend_opts)
def _nabla_z(
    pp : gtscript.Field[dtype],
    pz : gtscript.Field[dtype],
    dz : dtype
    ):

    with computation(PARALLEL), interval(...):
        dz2i = 0.5 / dz
        pz[0,0,0] = (pp[0,0,1] - pp[0,0,-1]) * dz2i
```
Note: the stencil is only ever executed on a subset of the vertical domain.

Error message:

```
Traceback (most recent call last):
  File "/home/tille/Development/FVM_GT4Py/tests/schaer_mountain.py", line 7, in <module>
    import fvms.model.config
  File "/home/tille/Development/FVM_GT4Py/src/fvms/model/config.py", line 15, in <module>
    from fvms.geometry.curvilinear import Metric
  File "/home/tille/Development/FVM_GT4Py/src/fvms/geometry/curvilinear.py", line 12, in <module>
    from fvms.operators.nabla import Nabla
  File "/home/tille/Development/FVM_GT4Py/src/fvms/operators/nabla.py", line 87, in <module>
    def _nabla_z(
  File "/home/tille/Development/FVM_GT4Py/src/fvms/utils/stencil.py", line 99, in _decorator
    stencil_obj = gt.gtscript.stencil(backend, definition, **kwargs)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gt4py/gtscript.py", line 266, in stencil
    return _decorator(definition)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gt4py/gtscript.py", line 254, in _decorator
    out = gt_loader.gtscript_loader(
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gt4py/loader.py", line 72, in gtscript_loader
    stencil_class = load_stencil("gtscript", backend, definition_func, externals, build_options)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gt4py/loader.py", line 58, in load_stencil
    return builder.build()
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gt4py/stencil_builder.py", line 72, in build
    stencil_class = self.backend.generate()
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gt4py/backend/gtc_backend/gtcpp/backend.py", line 230, in generate
    pyext_module_name, pyext_file_path = self.generate_extension()
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gt4py/backend/gtc_backend/gtcpp/backend.py", line 255, in generate_extension
    return super()._generate_extension(uses_cuda=False)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gt4py/backend/gtc_backend/gtcpp/backend.py", line 217, in _generate_extension
    return self.make_extension(gt_version=2, ir=self.builder.definition_ir, uses_cuda=uses_cuda)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gt4py/backend/gt_backends.py", line 613, in make_extension
    gt_pyext_sources: Dict[str, Any] = self.make_extension_sources(ir=ir)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gt4py/backend/gt_backends.py", line 649, in make_extension_sources
    gt_pyext_sources = gt_pyext_generator(ir)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gt4py/backend/gtc_backend/gtcpp/backend.py", line 72, in __call__
    oir = self._optimize_oir(oir)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gt4py/backend/gtc_backend/gtcpp/backend.py", line 85, in _optimize_oir
    oir = optimize_horizontal_executions(oir, GraphMerging)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gtc/passes/oir_dace_optimizations/api.py", line 28, in optimize_horizontal_executions
    sdfg = OirSDFGBuilder().visit(stencil)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/eve/visitors.py", line 121, in visit
    return visitor(node, **kwargs)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gtc/dace/oir_to_dace.py", line 632, in visit_Stencil
    return StencilOirSDFGBuilder.build(node.name, node, library_nodes)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/gtc/dace/oir_to_dace.py", line 381, in build
    res.validate()
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/dace/sdfg/sdfg.py", line 1928, in validate
    validate_sdfg(self)
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/dace/sdfg/validation.py", line 108, in validate_sdfg
    validate_state(start_state, sdfg.node_id(start_state), sdfg,
  File "/home/tille/Development/FVM_GT4Py/venv/lib/python3.9/site-packages/dace/sdfg/validation.py", line 428, in validate_state
    raise InvalidSDFGEdgeError(
dace.sdfg.validation.InvalidSDFGEdgeError: Memlet subset negative out-of-bounds (at state _nabla_z_state, edge pp[0:__I, 0:__J, -1:__K + 1] (pp:None -> VerticalLoop_140507677759664:IN_pp))
```

**Comment by gronerl** [comment]
So, you read outside of the K range, which I was told we shouldn't support when writing the gtc:dace backend / is undefined behavior. (There is also no checks for proper extents etc in  __call__ like for the other dimensions, nor does it show in the required halo etc. Since this is apparently never the less used in all user codes, I shall never the less try to add it in the cycle.

**Comment by gronerl** [comment]
So as a way to work around this is to make the interval(1, -1) and in turn pass 2 extra lines at top and bottom with the fields.

**Comment by tehrengruber** [comment]
This is exactly the workaround I added, but since it only occurs at a small number of cases and resizing the fields is no problem in FVM (just a one line change) I can live with the workaround for a while. Indeed we should fix this, after all a nabla operator is a common use case I guess.

**Comment by rheacangeo** [comment]
This seems to work now, is it fixed?

**Comment by gronerl** [comment]
It shouldn't work yet, but will be fixed with #468.
